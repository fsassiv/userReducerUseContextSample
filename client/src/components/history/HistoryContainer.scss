.history {
  min-height: 200px;
  width: 50%;
  &__top {
    margin-bottom: 10px;
  }
  &__tab {
    &-btn {
      display: inline-block;
      margin-left: 10px;
      &.active {
        font-weight: bold;
      }
    }
  }
}
