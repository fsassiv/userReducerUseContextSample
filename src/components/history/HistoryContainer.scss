.history {
  min-height: 200px;
  width: 50%;
  &__top {
  }
  &__tab {
    &-btn {
      display: inline-block;
      margin-left: 10px;
      &.active {
        font-weight: bold;
      }
    }
    &-wrapper {
      padding-top: 10px;
    }
  }
}
