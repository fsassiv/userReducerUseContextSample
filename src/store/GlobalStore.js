import React from "react";
import UserContextComponent from "./UserContext";
import HistoryContextComponent from "./HistoryContext";

const GlobalStore = ({ children }) => (
  <UserContextComponent>
    <HistoryContextComponent>{children}</HistoryContextComponent>
  </UserContextComponent>
);

export default GlobalStore;
