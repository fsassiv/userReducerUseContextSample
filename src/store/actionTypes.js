/*
Para grandes aplicações o mais recomendavel é que cada action create esteja em seu próprio .js
Ex: userActionTypes.js, historyActionType.js
*/

export const UserActionTypes = {
  SET_CURRENT_USER: "SET_CURRENT_USER",
  SIGN_UP_NEW_USER: "SIGN_UP_NEW_USER",
  LOGIN_USER: "LOGIN_USER",
  LOGOUT_USER: "LOGOUT_USER"
};

export const HistoryActionTypes = {
  SET_USER_ARTIST: "SET_USER_ARTIST",
  SET_USER_HISTORY:"SET_USER_HISTORY"
};
