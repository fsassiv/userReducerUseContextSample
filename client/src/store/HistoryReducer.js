import { HistoryActionTypes } from "./actionTypes";

export const INITIAL_STATE = {
  history: [],
  error: null
};

//copia do userReducer, somente para testar nested Context
export const historyReducer = (state, action) => {
  switch (action.type) {
    case HistoryActionTypes.SET_USER_HISTORY:
      return { ...state, history: action.payload };
    default:
      return state;
  }
};
