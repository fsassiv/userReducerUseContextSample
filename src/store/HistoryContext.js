import React, { useReducer, createContext } from "react";
import { INITIAL_STATE, historyReducer } from "./HistoryReducer";

export const HistoryContext = createContext();

const HistoryContextComponent = ({ children }) => {
  const [historyState, historyDispatch] = useReducer(
    historyReducer,
    INITIAL_STATE
  );

  return (
    <HistoryContext.Provider value={{ historyState, historyDispatch }}>
      {children}
    </HistoryContext.Provider>
  );
};

export default HistoryContextComponent;
