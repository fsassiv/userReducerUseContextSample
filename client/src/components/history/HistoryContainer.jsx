import React, { useState, useContext, useEffect } from "react";
import "./HistoryContainer.scss";
import { HistoryContext } from "../../store/HistoryContext";
import HistoryList from "./HistoryList";

const HistoryContainer = props => {
  const { historyState, historyDispatch } = useContext(HistoryContext);
  const [listSrc, setListSrc] = useState({});
  const [listType, setListType] = useState("artist");

  const handleTabBtnClick = event => {
    event.preventDefault();

    if (event.target.id !== "tab-one-btn") {
      setListType("album");
    } else {
      setListType("artist");
    }
  };

  useEffect(() => {
    if (historyState.history !== null) {
      if (listType === "artist") {
        setListSrc({ ...historyState.history });
        return;
      }
      //fallback to
      setListSrc({ ...historyState.history });
    }
  }, [historyState]);

  return (
    <div className="history">
      <div className="history__top">
        Últimos pesquisados:
        <a
          href="#"
          id="tab-one-btn"
          onClick={handleTabBtnClick}
          className={`history__tab-btn ${listType === "artist" && "active"}`}
        >
          Artistas
        </a>
        <a
          href="#"
          id="tab-two-btn"
          onClick={handleTabBtnClick}
          className={`history__tab-btn ${listType === "album" && "active"}`}
        >
          Álbuns
        </a>
      </div>
      {listSrc[listType] && (
        <HistoryList listType={listType} listSrc={listSrc[listType]} />
      )}
      {/* {!(Object.entries(listSrc).length>0 && listSrc[listType].length>0)?<p>Você ainda não possui históricos de busca...</p>:""} */}
    </div>
  );
};

export default HistoryContainer;
