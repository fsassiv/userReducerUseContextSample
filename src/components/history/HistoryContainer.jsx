import React, { useState, useContext, useEffect } from "react";
import "./HistoryContainer.scss";
import { HistoryActionTypes } from "../../store/actionTypes";
import { HistoryContext } from "../../store/HistoryContext";
// import HistoryArtistList from "./HistoryArtistList";
// import HistoryAlbumList from "./HistoryAlbumList";
import HistoryList from "./HistoryList";

const HistoryContainer = props => {
  const { historyState, historyDispatch } = useContext(HistoryContext);
  const [listSrc, setListSrc] = useState({});
  const [listType, setListType] = useState("artist");

  const handleTabBtnClick = event => {
    event.preventDefault();

    if (event.target.id !== "tab-one-btn") {
      setListType("album");
    } else {
      setListType("artist");
    }
  };

  useEffect(() => {
    if (historyState.history !== null) {
      if (listType === "artist") {
        setListSrc({ ...historyState.history });
        return;
      }
      setListSrc({ ...historyState.history });
    }
  }, [historyState]);

  return (
    <div className="history">
      <div className="history__top">
        Últimos pesquisados:
        <a
          href="#"
          id="tab-one-btn"
          onClick={handleTabBtnClick}
          className={`history__tab-btn ${listType === "artist" && "active"}`}
        >
          Artistas
        </a>
        <a
          href="#"
          id="tab-two-btn"
          onClick={handleTabBtnClick}
          className={`history__tab-btn ${listType === "album" && "active"}`}
        >
          Álbuns
        </a>
      </div>
      {listSrc[listType] && (
        <HistoryList listType={listType} listSrc={listSrc[listType]} />
      )}
      {!(Object.entries(listSrc).length>0 && listSrc[listType].length>0)?<p>Você ainda não possui históricos de busca...</p>:""}
    </div>
  );
};

export default HistoryContainer;
