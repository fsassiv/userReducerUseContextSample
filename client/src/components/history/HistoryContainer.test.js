import HistoryContainer from "./HistoryContainer";
import { HistoryContext } from "./../../store/HistoryContext";

describe("HistoryContainer", () => {
  const mockClick = jest.fn();

  const contextValues = {
    historyState: { history: [], error: null },
    historyDispatch: { type: "action.type", payload: "action.payload" }
  };

  const historyContainer = mount(
    <HistoryContext.Provider value={contextValues}>
      <HistoryContainer />
    </HistoryContext.Provider>
  );

  it("rendered correctly", () => {
    expect(historyContainer).toMatchSnapshot();
  });

  it("check if the element [history__tab-btn] exists", () => {
    expect(historyContainer.find("#tab-one-btn").exists()).toBe(true);
  });

  describe("click on 'history__tab-btn' [artist]", () => {
    beforeEach(() => {
      historyContainer.find("#tab-one-btn").simulate("click");
    });

    it("check click event", () => {
      expect(mockClick).toHaveBeenCalled();
    });
  });
});
