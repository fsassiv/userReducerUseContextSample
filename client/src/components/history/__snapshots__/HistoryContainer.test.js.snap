// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryContainer rendered correctly 1`] = `ReactWrapper {}`;
